Импорт библиотеки pandas для работы с табличными данными. Загрузка датасета movies.csv. Вывод первых 20 строк для просмотра.

In [5]:
import pandas as pd
df = pd.read_csv('movies.csv', sep =';')
df.head(n = 20)

title  release_year  score  \
0      David Attenborough: A Life on Our Planet        2020.0    9.0   
1                                     Inception        2010.0    8.8   
2                                  Forrest Gump        1994.0    8.8   
3                                    Anbe Sivam        2003.0    8.7   
4                            Bo Burnham: Inside        2021.0    8.7   
5                           Saving Private Ryan        1998.0    8.6   
6                              Django Unchained        2012.0    8.4   
7                                        Dangal        2016.0    8.4   
8                        Bo Burnham: Make Happy        2016.0    8.4   
9                         Louis C.K.: Hilarious        2010.0    8.4   
10              Dave Chappelle: Sticks & Stones        2019.0    8.4   
11                                     3 Idiots        2009.0    8.4   
12                                 Black Friday        2004.0    8.4   
13                                 Super Deluxe        2019.0    8.4   
14  Winter on Fire: Ukraine's Fight for Freedom        2015.0    8.3   
15                  Once Upon a Time in America        1984.0    8.3   
16                                  Taxi Driver        1976.0    8.3   
17                          Like Stars on Earth        2007.0    8.3   
18                            Bo Burnham: What.        2013.0    8.3   
19                            Full Metal Jacket        1987.0    8.3   

    number_of_votes  duration   main_genre main_production  
0           31180.0        83  documentary              GB  
1         2268288.0       148        scifi              GB  
2         1994599.0       142        drama              US  
3           20595.0       160       comedy              IN  
4           44074.0        87       comedy              US  
5         1346020.0       169        drama              US  
6         1472668.0       165      western              US  
7          180247.0       161       action              IN  
8           14356.0        60       comedy              US  
9           11973.0        84       comedy              US  
10          25687.0        65       comedy              US  
11         385782.0       170       comedy              IN  
12          20611.0       143        crime              IN  
13          13680.0       176     thriller              IN  
14          17710.0        98  documentary              UA  
15         342335.0       229        drama              US  
16         795222.0       113        crime              US  
17         188234.0       165        drama              IN  
18          11488.0        60       comedy              US  
19         723306.0       116        drama              GB

Описание предметной области данных – фильмы.  Вся информация этого датасета отображена в столбцах. В столбце "title" представлены наименования фильмов, в "release_year" - год выхода фильмов, в "score" - оценка фильма, в - "number_of_votes" количество голосов, в "duration" - длительность фильма, в "main_genre" - жанр фильма, в "main_production" - производство. \
С помощью метода info() осуществляется вывод информации о данных, а именно названия столбцов, типы данных, хранящиеся в них, и количество ненулевых значений в каждом столбце.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 389 entries, 0 to 388
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   title            384 non-null    object 
 1   release_year     387 non-null    float64
 2   score            386 non-null    float64
 3   number_of_votes  388 non-null    float64
 4   duration         389 non-null    int64  
 5   main_genre       389 non-null    object 
 6   main_production  388 non-null    object 
dtypes: float64(3), int64(1), object(3)
memory usage: 21.4+ KB


Выводы о данных. Столбцы "duration" и "main_genre" имеют 389 записей, в то время как столбец "title" - 384 записи, столбец "score" - 386, столбец "release_year" - 387, столбцы "number_of_votes" и "main_production" - 388 записи, что свидетельствует о наличии пропусков в датасете. Также метод info() выводит типы использованных данных, в текущем датасете это - odject, float64, int64. \
Затем с помощью метода describe() вывод статистической сводки каждого столбца, содержащего числовые значения.

In [9]:
df.describe()

release_year       score  number_of_votes    duration
count    387.000000  386.000000     3.880000e+02  389.000000
mean    2010.976744    7.512176     1.391523e+05  123.352185
std       10.805726    0.443870     2.365279e+05   28.304800
min     1954.000000    6.900000     1.013900e+04   28.000000
25%     2007.500000    7.100000     2.055375e+04  104.000000
50%     2014.000000    7.400000     4.606950e+04  122.000000
75%     2018.000000    7.800000     1.551585e+05  139.000000
max     2022.000000    9.000000     2.268288e+06  229.000000

Данные сильно сконцентрированы на современных фильмах. Медиана - 2014, то есть половина фильмов, приведенных в датасете, вышла позже 2014 года. Диапазон охватывает почти 70 лет, но акцент - на последние 10-15 лет. \
Распределение оценок сжато и смещено к верхней границе. Нет фильмов с низкими оценками (< 6.9), то есть это выборка лучших или отфильтрованных фильмов. \
Очень сильная положительная асимметрия по признаку количество голосов, вероятно, несколько сверхпопулярных фильмов тянут среднее вверх. Медиана почти в 3 раза меньше среднего, значит, большинство фильмов имеют умеренную популярность, а единицы - вирусную. \
Почти симметричное распределение по данным в столбце с продолжительностью - медиана ≈ среднее. Типичная длительность около 2 часов, что ожидаемо для фильмов. Минимальное значение 28 минут, скорее всего, это короткометражка. Максимум 229 минут, что вполне реально для фильмов 2000-ых. \
Определение корректности названий столбцов с помощью атрибута .columns.

In [10]:
df.columns

Index(['title', 'release_year', 'score', 'number_of_votes', 'duration',
       'main_genre', 'main_production'],
      dtype='object')

Наименования  корректны, а также правильно отражают суть столбцов. Все названия столбцов записаны одинаково - с маленькой буквы. Вместо пробелов используется нижнее подчеркивание. Тип данных используемый для названий столбцов - object. \
Проверка датасета на наличие пропусков с помощью метода isna(), подсчет пропусков с помощью метода sum().

In [11]:
df.isna()

title  release_year  score  number_of_votes  duration  main_genre  \
0    False         False  False            False     False       False   
1    False         False  False            False     False       False   
2    False         False  False            False     False       False   
3    False         False  False            False     False       False   
4    False         False  False            False     False       False   
..     ...           ...    ...              ...       ...         ...   
384  False         False  False            False     False       False   
385  False         False  False            False     False       False   
386  False         False  False            False     False       False   
387  False         False  False            False     False       False   
388  False         False  False            False     False       False   

     main_production  
0              False  
1              False  
2              False  
3              False  
4              False  
..               ...  
384            False  
385            False  
386            False  
387            False  
388            False  

[389 rows x 7 columns]

In [12]:
df.isna().sum()

title              5
release_year       2
score              3
number_of_votes    1
duration           0
main_genre         0
main_production    1
dtype: int64

Во всех колонках, кроме "duration" и "main_genre", есть пропуски. Заполнение пропусков. Для колонки "title" в качестве заменителя записывается – "Unknown film". Для колонки "main_production" – "Unknown production". Для числовых колонок заменителем будет выставлено среднее значение. Првторный вывод сводки по количеству пропусков в каждом столбце.

In [13]:
df['title'] = df['title'].fillna("Unknown film")
df['main_production'] = df['main_production'].fillna("Unknown production")
df['release_year'] = df['release_year'].fillna(df['release_year'].mean())
df['score'] = df['score'].fillna(df['score'].mean())
df['number_of_votes'] = df['number_of_votes'].fillna(df['number_of_votes'].mean())
df.isna().sum()

title              0
release_year       0
score              0
number_of_votes    0
duration           0
main_genre         0
main_production    0
dtype: int64

Все пропуски заполнены. \
Поиск явных дубликатов в данных, подсчет их количества с помощью методов .duplicated() и .sum().

In [ ]:
print(df.duplicated().sum())

2


Удаление дубликатов (drop_duplicates()) и обновление индексации (reset_index(drop=True)), чтобы избавиться от пропусков. Аргумент drop=True используется для того, чтобы при обновлении индексации не создался новый столбец с индексами.

In [ ]:
df = df.drop_duplicates().reset_index(drop=True)

Подсчет дубликатов для проверки.

In [ ]:
print(df.duplicated().sum())

0


Дубликаты не найдены. \
Поиск неявных дубликатов. Вывод списка уникальных значений в столбце "title", проверка наименований на уникальность и содержание ошибок. Замена некорректных названий на правильные (с помощью .replace()), повторная проверка дубликатов.

In [14]:
df['main_genre'].unique()


array(['documentary', 'scifi', 'drama', 'comedy', 'western', 'action',
       'crime', 'thriller', 'war', 'fantasy', 'romance', 'horror',
       'musical', 'animation', 'sports', 'COMEDY', 'dramaa'], dtype=object)

In [15]:
df['main_genre'] = df['main_genre'].replace('dramaa', 'drama')
df['main_genre'] = df['main_genre'].replace('COMEDY', 'comedy')
df['main_genre'].unique()

array(['documentary', 'scifi', 'drama', 'comedy', 'western', 'action',
       'crime', 'thriller', 'war', 'fantasy', 'romance', 'horror',
       'musical', 'animation', 'sports'], dtype=object)

Дубликаты не найдены. \
Проверка типов данных, вывод информацию о датасете при помощи метода info(). Приведение столбца release_year к числовому типу данных.

In [17]:
df = df.astype({"release_year" : 'int64'})
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 389 entries, 0 to 388
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   title            389 non-null    object 
 1   release_year     389 non-null    int64  
 2   score            389 non-null    float64
 3   number_of_votes  389 non-null    float64
 4   duration         389 non-null    int64  
 5   main_genre       389 non-null    object 
 6   main_production  389 non-null    object 
dtypes: float64(2), int64(2), object(3)
memory usage: 21.4+ KB
None


Для столбцов с числовыми данными определены числовые типы данных, для столбцов с текстовыми данными - object. \
Группировка по жанру и количеству фильмов каждого main_production. \
Для выполнения используются методы groupby() и agg().


In [20]:
grouped_df = df.groupby(['main_genre', 'main_production']).agg({'main_production': ['count']})
grouped_df.columns = ['count']
grouped_df

count
main_genre main_production       
action     DE                   1
           IN                   2
           US                   2
animation  JP                   1
           US                   2
...                           ...
thriller   TH                   1
           US                  21
war        CD                   1
           US                   2
western    US                   7

[94 rows x 1 columns]

Сортировка датафрейма по убыванию столбца 'count' с помощью метода sort_values(). Вывод отсортированного датафрейма.

In [21]:
sorted_df = grouped_df.sort_values(by='count', ascending=False)
sorted_df

count
main_genre main_production       
drama      US                  55
           IN                  48
comedy     US                  22
thriller   US                  21
           IN                  21
...                           ...
sports     DE                   1
thriller   IE                   1
           PS                   1
           TH                   1
war        CD                   1

[94 rows x 1 columns]

Больше всего фильмов снято в жанре drama. Больше всего фильмов в жанре drama снято производством US – 55. Меньшее количество снятых фильмов встречается в разных жанрах и производствах фильмов. \
Создается сводная таблица по средниему рейтингу фильмов по main_production. Сортировка по убыванию рейтинга.

In [22]:
data_pivot = pd.pivot_table(df, index='main_production', values='score', aggfunc='mean')
sorted_data_pivot = data_pivot.round(1).sort_values(by='score', ascending=False)
sorted_data_pivot

score
main_production          
CD                    8.2
ZA                    8.1
UA                    8.1
NZ                    7.8
TR                    7.8
MX                    7.7
IN                    7.6
AU                    7.6
TH                    7.6
MW                    7.6
GB                    7.6
ES                    7.6
US                    7.5
JP                    7.5
AR                    7.5
PS                    7.5
HK                    7.5
Unknown production    7.5
DE                    7.5
BE                    7.4
KR                    7.4
CA                    7.4
NO                    7.4
BR                    7.2
HU                    7.2
CN                    7.2
FR                    7.2
IT                    7.2
PL                    7.2
KH                    7.2
IE                    7.1
LT                    7.1
XX                    7.1
NL                    7.0
ID                    6.9
DK                    6.9

Как видно из результатов сортировки сводной таблицы, самый большой средний рейтинг у кинопроизводства CD (8,2), а самый маленький у ID (6,9). \
Создается сводная таблица по медианному рейтингу фильмов по жанрам и main_production. Сортировка по возрастанию (от A до Z) main_production.

In [24]:
data_pivot = pd.pivot_table(df, index='main_production', columns='main_genre', values='score', aggfunc='median')
sorted_data_pivot = data_pivot.sort_values(by='main_production', ascending=True)
sorted_data_pivot

main_genre          action  animation  comedy  crime  documentary     drama  \
main_production                                                               
AR                     NaN        NaN    7.50    NaN          NaN       NaN   
AU                     NaN        NaN    8.10    NaN          NaN  7.300000   
BE                     NaN        NaN     NaN    NaN          7.9       NaN   
BR                     NaN        NaN     NaN    NaN          7.2       NaN   
CA                     NaN        NaN    7.10    NaN          NaN  7.150000   
CD                     NaN        NaN     NaN    NaN          NaN       NaN   
CN                     NaN        NaN     NaN    NaN          NaN  7.250000   
DE                    7.10        NaN     NaN    NaN          NaN  7.150000   
DK                     NaN        NaN     NaN   6.90          NaN       NaN   
ES                     NaN        NaN    8.10    NaN          NaN  7.400000   
FR                     NaN        NaN    7.20    NaN          NaN  7.100000   
GB                     NaN        NaN    8.00   7.40          9.0  7.150000   
HK                     NaN        NaN     NaN    NaN          NaN  7.500000   
HU                     NaN        NaN     NaN    NaN          NaN  7.000000   
ID                     NaN        NaN     NaN    NaN          NaN       NaN   
IE                     NaN        NaN     NaN    NaN          NaN  7.000000   
IN                    8.25        NaN    7.60   7.40          NaN  7.600000   
IT                     NaN        NaN     NaN    NaN          NaN  7.000000   
JP                     NaN       7.10    7.60    NaN          NaN  7.400000   
KH                     NaN        NaN     NaN    NaN          NaN  7.200000   
KR                     NaN        NaN     NaN    NaN          NaN  7.650000   
LT                     NaN        NaN     NaN    NaN          NaN  7.100000   
MW                     NaN        NaN     NaN    NaN          NaN  7.600000   
MX                     NaN        NaN     NaN    NaN          NaN  7.700000   
NL                     NaN        NaN     NaN    NaN          NaN  7.050000   
NO                     NaN        NaN    6.90   7.80          NaN       NaN   
NZ                     NaN        NaN    7.80    NaN          NaN  7.800000   
PL                     NaN        NaN     NaN    NaN          NaN       NaN   
PS                     NaN        NaN     NaN    NaN          NaN       NaN   
TH                     NaN        NaN     NaN    NaN          NaN       NaN   
TR                     NaN        NaN    7.85    NaN          NaN  7.750000   
UA                     NaN        NaN     NaN    NaN          8.1       NaN   
US                    7.45       7.25    7.30   7.35          7.6  7.500000   
Unknown production     NaN        NaN     NaN    NaN          NaN  7.512176   
XX                     NaN        NaN    7.10    NaN          NaN       NaN   
ZA                     NaN        NaN     NaN    NaN          8.1       NaN   

main_genre          fantasy  horror  musical  romance  scifi  sports  \
main_production                                                        
AR                      NaN     NaN      NaN      NaN    NaN     NaN   
AU                      NaN     NaN      NaN      NaN    NaN     NaN   
BE                      NaN    7.00      NaN      NaN    NaN     NaN   
BR                      NaN     NaN      NaN      NaN    NaN     NaN   
CA                      NaN     NaN      NaN      NaN   8.00     NaN   
CD                      NaN     NaN      NaN      NaN    NaN     NaN   
CN                     7.20     NaN      NaN      NaN    NaN     NaN   
DE                      NaN    7.50      NaN      NaN    NaN     7.4   
DK                      NaN     NaN      NaN      NaN    NaN     NaN   
ES                     7.50    7.00      NaN      NaN    NaN     NaN   
FR                     7.50     NaN      NaN      NaN    NaN     NaN   
GB                      NaN    7.10      NaN      7.3   7.95 

Исходя из данных полученной таблицы можно увидеть медианные оценки фильмов в различных жанрах всех производств, отсортированных в порядке возрастания по производству. Например, медианная оценка производства фильмов AR в жанре comedy составляет 7.5. Медианная оценка производства фильмов KR в жанре triller составляет 7.25.


Выделяется топ 7 жанров с самым высоким рейтингом. Так как к одному жанру может принадлежать несколько фильмов, соответственно, несколько оценок, рассматриваются все жанры и  средний рейтинг по каждому жанру. Используется группировка данных по жанрам. Для каждого жанра вычисляется средний рейтинг с помощью .mean().

In [25]:
mean_score = df.groupby('main_genre')['score'].mean()
mean_score

main_genre
action         7.700000
animation      7.200000
comedy         7.608621
crime          7.466667
documentary    7.760000
drama          7.478969
fantasy        7.484211
horror         7.350000
musical        7.100000
romance        7.405342
scifi          7.771429
sports         7.400000
thriller       7.483607
war            7.800000
western        7.557143
Name: score, dtype: float64

Сортировка в порядке убывания полученного рейтинга жанров. Срез по индексу с помощью .index[:7]. Результат top_genres - список топ 7 жанров с самым высоким средним рейтингом.

In [26]:
mean_score_sorted = mean_score.round(2).sort_values(ascending=False)
top_genres = mean_score_sorted.index[:7]
top_genres

Index(['war', 'scifi', 'documentary', 'action', 'comedy', 'western',
       'fantasy'],
      dtype='object', name='main_genre')

Фильтрация датафрейма в соответсвии с полученным топом жанров. Результат фильтрации - датафрейм, содержащий фильмы по 7 жанрам.

In [27]:
filtered_df = df[df['main_genre'].isin(top_genres)]
filtered_df

title  release_year  score  \
0         David Attenborough: A Life on Our Planet          2020    9.0   
1                                        Inception          2010    8.8   
3                                       Anbe Sivam          2003    8.7   
4                               Bo Burnham: Inside          2021    8.7   
6                                 Django Unchained          2012    8.4   
..                                             ...           ...    ...   
373                                       2 States          2014    6.9   
375                                 The Half of It          2020    6.9   
379  Legend of the Guardians: The Owls of Ga'Hoole          2010    6.9   
382                                       The Trip          2021    6.9   
386                            Sorry to Bother You          2018    6.9   

     number_of_votes  duration   main_genre main_production  
0            31180.0        83  documentary              GB  
1          2268288.0       148        scifi              GB  
3            20595.0       160       comedy              IN  
4            44074.0        87       comedy              US  
6          1472668.0       165      western              US  
..               ...       ...          ...             ...  
373          25344.0       149       comedy              IN  
375          34959.0       104       comedy              US  
379          82623.0       100      fantasy              US  
382          19706.0       113       comedy              NO  
386          75653.0       111      fantasy              US  

[119 rows x 7 columns]

Выборка топ 10 лет по количеству записей. Для этого используется функция .value_counts(). Результат - отсортированный массив с перечнем годов и соответсвующим количеством фильмов.

In [28]:
year_counts = filtered_df['release_year'].value_counts()
year_counts

release_year
2019    13
2018    10
2021    10
2017    10
2016     9
2014     8
2015     6
2010     6
2020     5
2011     4
2006     4
2012     4
2013     3
2007     3
2004     2
1997     2
2009     2
1979     2
2008     2
1975     1
2003     1
2001     1
1982     1
1987     1
1994     1
1961     1
2005     1
1966     1
2022     1
1998     1
1996     1
1980     1
1991     1
Name: count, dtype: int64

Срез первых 10 лет, где количество фильмов максимальное. Результат - список топ 10 лет по количеству фильмов.

In [29]:
top_years = year_counts.index[:10]
top_years

Index([2019, 2018, 2021, 2017, 2016, 2014, 2015, 2010, 2020, 2011], dtype='int64', name='release_year')

Фильтрация датафрейма в соответсвии с полученным топом жанров. Результат фильтрации - датафрейм с фильмами из топ 7 жанров и топ 10 лет.

In [30]:
filtered_df_by_year_and_genre = filtered_df[filtered_df['release_year'].isin(top_years)]
filtered_df_by_year_and_genre

title  release_year  score  \
0         David Attenborough: A Life on Our Planet          2020    9.0   
1                                        Inception          2010    8.8   
4                               Bo Burnham: Inside          2021    8.7   
7                                           Dangal          2016    8.4   
8                           Bo Burnham: Make Happy          2016    8.4   
..                                             ...           ...    ...   
373                                       2 States          2014    6.9   
375                                 The Half of It          2020    6.9   
379  Legend of the Guardians: The Owls of Ga'Hoole          2010    6.9   
382                                       The Trip          2021    6.9   
386                            Sorry to Bother You          2018    6.9   

     number_of_votes  duration   main_genre main_production  
0            31180.0        83  documentary              GB  
1          2268288.0       148        scifi              GB  
4            44074.0        87       comedy              US  
7           180247.0       161       action              IN  
8            14356.0        60       comedy              US  
..               ...       ...          ...             ...  
373          25344.0       149       comedy              IN  
375          34959.0       104       comedy              US  
379          82623.0       100      fantasy              US  
382          19706.0       113       comedy              NO  
386          75653.0       111      fantasy              US  

[81 rows x 7 columns]

Создание сводной таблицы. В качестве аргументов сводной таблицы передается 'release_year' в строки (выборка топ 10 лет), 'main_genre' в столбцы (выборка топ 7 жанров), 'duration' в значение.

In [31]:
pivot_table = pd.pivot_table(filtered_df_by_year_and_genre, index='release_year', columns='main_genre', values='duration', aggfunc='median')
pivot_table

main_genre    action  comedy  documentary  fantasy  scifi    war  western
release_year                                                             
2010             NaN    99.5          NaN     99.0  148.0    NaN    110.0
2011           139.0   134.0          NaN     94.0    NaN    NaN      NaN
2014             NaN   125.5         82.5    102.0    NaN   90.0      NaN
2015            28.0   123.0         98.0      NaN    NaN  137.0    188.0
2016           161.0   101.0        100.0    108.0    NaN    NaN    102.0
2017             NaN   114.0         97.0    142.0  164.0    NaN      NaN
2018             NaN    79.0         98.0    107.5   90.0    NaN    132.0
2019             NaN    95.0        104.0     81.0    NaN    NaN      NaN
2020             NaN   104.0         84.5      NaN    NaN    NaN      NaN
2021           158.0   100.0         90.5     91.0  138.0    NaN      NaN

Эта сводная таблица представляет медианное значение длительности фильмов в минутах для разных жанров и лет (с 2010 по 2021 годы) по семи жанрам: action, comedy, documentary, drama, scifi, war и western. Для некоторых жанров и лет отсутствуют данные. Например, для жанров action, documentary, scifi, war и western часто встречаются пропуски, вследствии недостаточного количества фильмов этих жанров в рассматриваемых годах для расчета медианы. Жанры comedy и drama присутствуют почти во всех годах. Это может указывать на их высокую популярность и стабильное производство фильмов этих жанров.
Сводная таблица показывает, что более популярные жанры, такие как comedy и drama, имеют данные по большинству лет, в то время как жанры, такие как action, scifi, war и western, представлены с пропусками. Некоторые жанры (например, scifi, western) часто имеют высокие медианные значения, что свидетельствует о том, что они склонны к созданию более длинных фильмов.